rfc-128-continuous-deployment.md

```diff
@@ -0,0 +1,223 @@
+
+### Safety concerns
+
+The person who merges a PR must be confident that it is safe to do so. As part of the trial of CD, [we added PR templates to each affected app](https://github.com/alphagov/smart-answers/pull/4758), to explicitly warn developers about the automatic deployment.
+
+However, we also want to ensure that a faulty PR merged in haste can be caught before it impacts the live site. There are a number of checks we can perform to help ensure this, without aiming for perfection.
+
+> **We won't check alerts.** At the time of writing, we have no evidence that checking Icinga alerts has any advantage over the checks discussed below. We can investigate such a check if we find evidence [[1](https://github.com/alphagov/govuk-puppet/pull/10597)].
+
+> **We will only consider _live_ code.** Interactions due to manual and scheduled tasks can be checked manually, using a branch deployment. In the worst case, we should have ample time to fix any faults that develop.
+
+#### Check: Internal repo tests pass (i.e. features are tested in detail)
+
+For the purpose of automating deployments:
+
+- We will use code coverage as our indicator of previous, thorough testing. While an app may have a [rigorous testing strategy](https://github.com/alphagov/content-publisher/blob/69bed3535de1e65bda9d5909a3bc56a319e46576/docs/testing-strategy.md), it's hard to know if this has been applied consistently as the app has evolved.
+
+- We will require a minimum level of coverage that is realistic for all of our apps. We will not pursue this once CD is enabled, as doing so could encourage [a lack of rigour in the way we write our tests](https://jeroenmols.com/blog/2017/11/28/coveragproblem/).
+
+- We will not check coverage of periphery code e.g. CSS, JavaScript. This is because it has a low footprint in our codebases, and it's hard to measure coverage for: the frameworks we use don't support it.
+
+An app has "enough" of these tests when:
+
+- Its code coverage exceeds 95% at the point when CD is enabled.
+
+#### Check: API [contract tests](https://docs.publishing.service.gov.uk/manual/pact-broker.html) pass (i.e. adapters work with their APIs)
+
+For the purpose of automating deployments:
+
+- We only need to add contract tests when it's hard to be confident about an API change. If an API endpoint only has a single "consumer" app, then it's easy to gain confidence by manually testing the pair still work together.
+
+- We only need to add contract tests for the APIs we consume internally. Some of our APIs may be public, but this is a separate, product-level commitment. Our strategy for testing public APIs is out-of-scope for this RFC.
+
+With this in mind, an API app has "enough" of these tests when:
+
+- Each endpoint with multiple consumers has at least one contract test.
```

Comment:
There a few apps with APIs that only have a single consumer (that [I can find](https://github.com/alphagov/govuk-rfcs/blob/continuous-deployment/rfc-128/e2e-interactions.md)):

- Collections
- Content Data API
- Imminence
- Maslow
- Router API
- Support
- Support API

Part of the motivation behind the "rule of 2" is that I didn't want to prematurely harden our APIs with contract tests. For me, the existence of a contract test should be the indicator that an API is stable and open for reuse, and I'd be suspicious of using an API otherwise. I think this applies well to Support*, Router and Content Data APIs, which are primarily a way of splitting code between two repos, rather than a general facility for any GOV.UK app to make reliable use of.

I don't think we should bother doing anything for Maslow.

However, the Imminence and Collections APIs are harder to justify, because it's not immediately apparent what consumers they have, and the general nature of their APIs means they _could_ have more consumers in future. I can't think of a simple rule that would include these. However, would could agree they are a special case and _must_ have contracts tests as a criteria for enabling CD. How does that sound?

- @sihugh and @thomasleese can you comment on these apps, since you've worked on them? Are we significantly concerned about automating deployments for them without contract tests? Is it worth the effort?

Reply:
I'm not very familiar with Imminence, but a few of our apps do use the APIs provided by Collections ([Signon](https://github.com/alphagov/signon/blob/9aa4910b518ea22f9e190ab50fec138108c0de63/lib/organisations_fetcher.rb#L24), [Transition](https://github.com/alphagov/transition/blob/9d44d66b5a157b5089c53395ac74dab3ce2c96bc/lib/transition/import/whitehall_orgs.rb#L51-L52)) so I would expect that we'll need contract tests for them. I don't see that as a special case, because clearly here the API has multiple consumers.

I'm sure we'll miss some APIs that have multiple consumers that we're not aware of, but I think we can deal with that as and when it comes up, and add contract tests when we learn about it.

Reply:
Thanks @thomasleese, that's really helpful. With Collections, I did notice those API calls when I did my audit; neither of them are from live code, though, so they didn't contribute towards the total count.

However, it sounds like we have some consensus that Collections should be a special case and contract tests will be a requirement for enabling CD for it. @thomasleese @sihugh @bevanloon do you all agree?

@sihugh did you have an opinion on Imminence? I'm not fussed either way, it's just that we need to make a choice of how much effort we're committing to for each app, if we consider it to be a special case.

Reply:
> Part of the motivation behind the "rule of 2" is that I didn't want to prematurely harden our APIs with contract tests. For me, the existence of a contract test should be the indicator that an API is stable and open for reuse, and I'd be suspicious of using an API otherwise. I think this applies well to Support*, Router and Content Data APIs, which are primarily a way of splitting code between two repos, rather than a general facility for any GOV.UK app to make reliable use of.

Perhaps that hardening would be a de-motivating aspect to creating new distributed apps 😁  A lot of the smells in our architecture seem to be cases where only one app speaks to another. 

> - Support
> - Support API

These are both apps that act as APIs? 😬 

--- 

I wonder if this is an aspect of legacy and future maybe. I can definitely understand the pragmatism of not wanting to do contract tests for all APIs but then if I was making something new it'd feel weird to not consider whether to add contract tests until I had a certain number of consumers. Perhaps there's a middle ground all new APIs should have them?

Reply:
Hmm, it does seem quite a difficult one to argue for. I think it's reasonable for people to want to split large masses of functionality somewhere, and there are benefits to having such a distinct interface boundaries (forward planning, clear abstraction). However, if we look at what we've actually got:

- Support/API makes no sense to me, architecturally.
- Router/API also makes no sense to me (with it's weird "reload" feature).
- Maslow is not worth commenting on.
- Collections is a good example of a need for contract tests.
- Imminence also seems worthy of some contract tests.
- Content Data API is a less clear example.

Looking at it that way, I think there's a good case for future apps having contract tests, as half the apps we proposing as exempt are verging on nonsensical, architecturally.

@theseanything as someone who helped created one of the most recent APIs for GOV.UK, what are your thoughts on having contract tests for new apps - should they be required?

Reply:
Yup - so for Content Data API, contract testing wasn't seen as a feasible testing strategy. At the time, the long term plan for Content Data API was that anyone/thing could use the API (i.e. consumers we weren't responsible for the development of), not just the frontend application. [Here's a good list of conditions when contract testing doesn't work](https://docs.pact.io/getting_started/what_is_pact_good_for). 

I don't think there should be a concrete requirement for contract testing on every app (old or new) as its not suitable for every application.

Sorry its not very helpful - but I'm not sure what blanket rules we can put in place to cover all our APIs for adequate testing. Maybe there are other testing strategies that sufficient? I.e. Making sure we have feature/integrations tests that cover all endpoints and responses. 

(Also a little unclear what we mean by "hardening an API"?)

---

#### Bonus context (- not necessarily relevant for this RFC)
Our alternative approach (for some of the better architected endpoints), we made an effort to document the API and add feature tests (request specs) for guarantees we made in the [documentation](https://github.com/alphagov/content-data-api/tree/master/doc/api). We also (partly, but not very well) versioned our API and served multiple versions. This meant that we could decouple the API from the frontend application - i.e. if we changed the API, the change was documented, versioned and made live, but we didn't have to force the frontend into using it straight away nor did we break existing functionality as it just requested the older compatible version of the API. This worked pretty well for us!

Reply:
Thanks both 🚀. I've added another commit - https://github.com/alphagov/govuk-rfcs/pull/128/commits/ac9814b0febf5f8d120bbb38fbc0eed7c8cbc4f9 - and will resolve this now.